sql/core/src/main/scala/org/apache/spark/sql/json/JsonRDD2.scala

```diff
@@ -0,0 +1,409 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.json
+
+import java.io.ByteArrayOutputStream
+import java.sql.Timestamp
+
+import scala.collection.Map
+
+import com.fasterxml.jackson.core._
+
+import org.apache.spark.rdd.RDD
+import org.apache.spark.sql.catalyst.analysis.HiveTypeCoercion
+import org.apache.spark.sql.catalyst.expressions._
+import org.apache.spark.sql.types._
+import org.apache.spark.Logging
+
+private[sql] object JsonRDD2 extends Logging {
+  def jsonStringToRow(
+      json: RDD[String],
+      schema: StructType,
+      columnNameOfCorruptRecords: String): RDD[Row] = {
+    parseJson(json, schema, columnNameOfCorruptRecords)
+  }
+
+  def inferSchema(
+      json: RDD[String],
+      samplingRatio: Double = 1.0,
+      columnNameOfCorruptRecords: String): StructType = {
+    require(samplingRatio > 0, s"samplingRatio ($samplingRatio) should be greater than 0")
+    val schemaData = if (samplingRatio > 0.99) {
+      json
+    } else {
+      json.sample(withReplacement = false, samplingRatio, 1)
+    }
+
+    // perform schema inference on each row and merge afterwards
+    schemaData.mapPartitions { iter =>
+      val factory = new JsonFactory()
+      iter.map { row =>
+        try {
+          val parser = factory.createParser(row)
+          parser.nextToken()
+          inferField(parser)
+        } catch {
+          case _: JsonParseException =>
+            StructType(Seq(StructField(columnNameOfCorruptRecords, StringType)))
+        }
+      }
+    }.treeAggregate[DataType](StructType(Seq()))(compatibleRootType, compatibleRootType) match {
+      case st: StructType => nullTypeToStringType(st)
+    }
+  }
+
+  /**
+   * Infer the type of a json document from the parser's token stream
+   */
+  private def inferField(parser: JsonParser): DataType = {
+    import com.fasterxml.jackson.core.JsonToken._
+    parser.getCurrentToken match {
+      case null | VALUE_NULL => NullType
+
+      case FIELD_NAME =>
+        parser.nextToken()
+        inferField(parser)
+
+      case VALUE_STRING if parser.getTextLength < 1 => NullType
```

Comment:
Yes, an empty string gets inferred as a NullType. After inference is
complete any remaining NullType fields get converted back to a StringType.
The old code does this and has test coverage for it, but it does seem a bit
odd.
On May 3, 2015 9:03 PM, "Yin Huai" <notifications@github.com> wrote:

> In sql/core/src/main/scala/org/apache/spark/sql/json/JsonRDD2.scala
> <https://github.com/apache/spark/pull/5801#discussion_r29565168>:
>
> > +    }
> > +  }
> > +
> > +  /**
> > +   * Infer the type of a json document from the parser's token stream
> > +   */
> > +  private def inferField(parser: JsonParser): DataType = {
> > +    import com.fasterxml.jackson.core.JsonToken._
> > +    parser.getCurrentToken match {
> > +      case null | VALUE_NULL => NullType
> > +
> > +      case FIELD_NAME =>
> > +        parser.nextToken()
> > +        inferField(parser)
> > +
> > +      case VALUE_STRING if parser.getTextLength < 1 => NullType
>
> Does it mean that we get an empty string? If so, can we keep the
> StringType? Otherwise, I feel we are destroying information.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/apache/spark/pull/5801/files#r29565168>.
>